2.0-taproot-introduction.ipynb

```diff
@@ -0,0 +1,354 @@
+    "\n",
+    "This sequence of steps is illustrated below:"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "![test](images/segwit_version1_1.jpg)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "In each chapter, we'll implement the spending policy using both v0 (pre-taproot) segwit and v1 (taproot) segwit outputs, and highlight the differences in transaction weight and privacy."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Transaction sequence\n",
+    "\n",
+    "This chapter demonstrates the transaction sequence in full detail. Future chapters follow the same steps, but use convenience functions to abstract away the low-level details."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example 2.0.1 Generate a segwit v0 bech32 address\n",
+    "\n",
+    "In this example, we'll use a v0 (pre-taproot) segwit output."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Generate a new key pair\n",
+    "privkey, pubkey = generate_key_pair()\n",
+    "print(\"Pubkey: {}\\n\".format(pubkey.get_bytes().hex()))\n",
+    "\n",
+    "# Get the hash160 of the public key for the witness program\n",
+    "program = hash160(pubkey.get_bytes())\n",
+    "print(\"Witness program: {}\\n\".format(program.hex()))\n",
+    "\n",
+    "# Create (regtest) bech32 address\n",
+    "version = 0x00\n",
+    "address = program_to_witness(version, program)\n",
+    "print(\"bech32 address: {}\".format(address))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### Example 2.0.2 Start a Bitcoin Core node, then generate blocks and send output to the bech32 address generated above\n",
+    "\n",
+    "This functionality will be encapsulated in the `generate_and_send_coins(node, address)` method later."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Start node\n",
+    "test = util.TestWrapper()\n",
+    "test.setup(num_nodes=1)\n",
+    "\n",
+    "version = test.nodes[0].getnetworkinfo()['subversion']\n",
+    "print(\"\\nClient version is {}\\n\".format(version))\n",
+    "\n",
+    "# Generate 101 blocks\n",
+    "test.nodes[0].generate(101)\n",
+    "balance = test.nodes[0].getbalance()\n",
+    "print(\"Balance: {}\\n\".format(balance))\n",
+    "\n",
+    "assert balance > 1\n",
+    "\n",
+    "unspent_txid = test.nodes[0].listunspent(1)[-1][\"txid\"]\n",
+    "inputs = [{\"txid\": unspent_txid, \"vout\": 0}]\n",
+    "\n",
+    "# Create a raw transaction sending 1 BTC to the address, then sign and send it.\n",
+    "# We won't create a change output, so maxfeerate must be set to 0\n",
```

Comment:
Suggestion: Mention that maxfeerate will be set a couple lines down at `test.nodes[0].sendrawtransaction(hexstring=tx_hex, maxfeerate=0)`, I wasn't initally didn't quite understand where this was set.

Reply:
fixed